qmp/libvirtgo_linux.go

```diff
@@ -0,0 +1,327 @@
+)
+
+// LibvirtGoEventsInterval constant used by the getPollInterval
+// method.
+const LibvirtGoEventsInterval = "LIBVIRTGO_EVENTS_INTERVAL"
+
+type libvirtGoMonitorLinux struct {
+	LibvirtGoMonitor
+	domainName string
+	domain     *libvirt.VirDomain
+	uri        string
+	virConn    *libvirt.VirConnection
+	once       sync.Once
+	doneChan   chan bool
+	eventsChan chan Event
+}
+
+// NewLibvirtGoMonitor configures a connection to the provided hypervisor
+// and domain.
+// An error is returned if the provided libvirt connection URI is invalid.
+//
+// Hypervisor URIs may be local or remote, e.g.,
+//	qemu:///system
+//	qemu+ssh://libvirt@example.com/system
+func NewLibvirtGoMonitor(uri, domain string) Monitor {
+	return &libvirtGoMonitorLinux{
+		uri:        uri,
+		domainName: domain,
+	}
+}
+
+// Connect  sets up QEMU QMP connection via libvirt using
+// the libvirt-go package.
+// An error is returned if the libvirtd daemon is unreachable.
+func (mon *libvirtGoMonitorLinux) Connect() error {
+
+	// As per the libvirt core library:
+	// For proper event handling, it is important
+	// that the event implementation is registered
+	// before a connection to the Hypervisor is opened.
+	// We only do this once the first a Connect is called.
+	mon.once.Do(eventRegisterDefaultImpl)
+
+	// Already connected?
+	if mon.virConn != nil {
+		return nil
+	}
+
+	virConn, err := newVirConnectionInternal(mon.uri)
+	if err != nil {
+		return err
+	}
+
+	domain, err := lookupDomainByName(&virConn, mon.domainName)
+	if err != nil {
+		closeConnectionInternal(&virConn)
+		return err
+	}
+
+	mon.domain = &domain
+	mon.virConn = &virConn
+	mon.eventsChan = make(chan Event)
+	mon.doneChan = make(chan bool)
+
+	return err
+}
+
+var newVirConnectionInternal = func(
+	uri string) (libvirt.VirConnection, error) {
+	return libvirt.NewVirConnection(uri)
+}
+
+// Disconnect tears down open QMP socket connections.
+func (mon *libvirtGoMonitorLinux) Disconnect() error {
+	var err error
+	if mon.domain != nil {
+		mon.domain.Free()
+		mon.domain = nil
+	}
+
+	if mon.virConn != nil {
+		_, err = closeConnectionInternal(mon.virConn)
+		mon.virConn = nil
+		close(mon.doneChan)   // stop listenning to events
+		close(mon.eventsChan) // stop sending events to clients
+	}
+
+	return err
+}
+
+var closeConnectionInternal = func(
+	virConn *libvirt.VirConnection) (int, error) {
+	return virConn.CloseConnection()
+}
+
+// Run executes the given QAPI command against a domain's QEMU instance.
+// For a list of available QAPI commands, see:
+//	http://git.qemu.org/?p=qemu.git;a=blob;f=qapi-schema.json;hb=HEAD
+func (mon *libvirtGoMonitorLinux) Run(cmd []byte) ([]byte, error) {
+	if mon.virConn == nil {
+		return nil, errors.New("Run() need a established connection to proceed.")
+	}
+
+	result, err := qemuMonitorCommandInternal(mon.domain, string(cmd))
+	if err != nil {
+		return nil, err
+	}
+
+	return []byte(result), nil
+}
+
+var qemuMonitorCommandInternal = func(
```

Comment:
Swapping out functions during tests is messy.  Make this an interface and make implementations of that for the monitor and the tests.